source/common/router/router.cc

```diff
@@ -487,7 +488,24 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
+  const auto http_protocol_options = cluster_->httpProtocolOptions();
+
+  if (http_protocol_options.has_value() && http_protocol_options.value().auto_sni()) {
+    const auto host_str = headers.Host()->value().getStringView();
+    const auto parsed_authority = Http::Utility::parseAuthority(host_str.data());
+
+    if (!parsed_authority.is_ip_address_) {
```

Comment:
Please add an integration test with IP address in the host header to make sure we don't set SNI in this case.